Bluetooth: Mesh: Split Proxy services

LingaoM opened the pull request:
we will no longer need the additional `disconnect` parameter,
such as we only process gatt database from disconnect handler.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Delete bt_mesh_proxy_prov_disable parameter.

we will no longer need the additional `disconnect` parameter,
such as we only process gatt database from disconnect handler.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik commented:
> This PR contains multiple commits, which will be used to split the proxy services

@LingaoM this this end up changing, or is this still a work in progress? Is it ready for review?

LingaoM commented:
I'm not ready. I'll deal with https://github.com/zephyrproject-rtos/zephyr/pull/36721#issuecomment-878286532 first, and then split the proxy. Maybe the following can be put in this pr.

LingaoM commented:
`gatt_svc` and `filter_type` will be processed in subsequent commits. The final is a complete split.

trond-snekvik commented on subsys/bluetooth/mesh/gatt_services.c:
```diff
@@ -91,6 +97,7 @@ static struct bt_mesh_proxy_client {
 		BLACKLIST,
 		PROV,
 	} filter_type;
+	uint8_t gatt_type;
```
This is in preparation for the `MESH_GATT_` enum to get a `MESH_GATT_PROXY_CLIENT` as well, right?

trond-snekvik commented on subsys/bluetooth/mesh/gatt_services.c:
```diff
@@ -740,20 +732,9 @@ int bt_mesh_proxy_gatt_disable(void)
-	gatt_svc = MESH_GATT_NONE;
-
-	return 0;
+	return bt_gatt_service_unregister(&proxy_svc);
```
Should this still set `client->gatt_type  = MESH_GATT_NONE`, or is this always handled by the disconnected callback?

trond-snekvik commented on subsys/bluetooth/mesh/gatt_services.c:
```diff
@@ -1017,46 +993,22 @@ int bt_mesh_proxy_prov_enable(void)
-
-	if (gatt_svc != MESH_GATT_PROV) {
-		return -EBUSY;
-	}
```
The removal of these checks everywhere makes me a bit uncomfortable. Are they redundant because of checks in the gatt subsystem?

trond-snekvik commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -640,7 +644,7 @@ int bt_mesh_prov_disable(bt_mesh_prov_bearer_t bearers)
 	if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT) &&
 	    (bearers & BT_MESH_PROV_GATT)) {
-		bt_mesh_proxy_prov_disable(true);
+		(void)bt_mesh_proxy_prov_disable();
```
Should also disconnect?

LingaoM replied:
Currentlly in pb_gatt disconnect cb --> bt_mesh_pb_gatt_close --> link_closed --> cb->link_closed --> reset_state() --> Clear LINK_ACTIVE.

And we add
`	if (bt_mesh_prov_active()) {
		return -EBUSY;
	}` before to avoid this

trond-snekvik commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -61,7 +61,7 @@ static int mesh_commit(void)
 
 	if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
-		bt_mesh_proxy_prov_disable(true);
+		(void)bt_mesh_proxy_prov_disable();
```
Should also disconnect?

LingaoM replied:
In this mesh_commit, are not all calls after power up? Will there be a connection at this time?

LingaoM committed:
Bluetooth: Mesh: Split gatt services to pb-gatt and proxy

Split gatt services to pb-gatt-srv and proxy-srv.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented on subsys/bluetooth/mesh/CMakeLists.txt:
```diff
@@ -39,9 +39,12 @@ zephyr_library_sources_ifdef(CONFIG_BT_MESH_PROVISIONER provisioner.c)
-zephyr_library_sources_ifdef(CONFIG_BT_MESH_PB_GATT pb_gatt.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MESH_PB_GATT
+    pb_gatt.c
+    pb_gatt_srv.c
```
@trond-snekvik Maybe we add `CONFIG_BT_MESH_PB_GATT_SRV` to replace `CONFIG_BT_MESH_PB_GATT`, and define another `CONFIG_BT_MESH_PB_GATT_CLI`, but it's best to solve it after the proxy client is implemented.

trond-snekvik replied:
It can probably wait until it makes sense. It's also an API change of sorts, so it might not be worth doing

LingaoM committed:
Bluetooth: Mesh: Rename bt_mesh_pb_gatt<*> function

Rename bt_mesh_pb_gatt_<*> function, and split from proxy.h

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
@trond-snekvik It should be reviewable now.

PavelVPV commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -369,10 +354,17 @@ int bt_mesh_start(void)
+	/* For PB-GATT provision, will enable in le disconnect handler. */
+	if (bt_mesh_prov_link.bearer->type == BT_MESH_PROV_ADV) {
+		if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
+			(void)bt_mesh_pb_gatt_disable();
```
If PB-GATT is disabled here, then PB-GATT service is unregistered before the provisioner disconnects from PB-GATT:
Provisioning Data PDU received -> `prov_device.c:prov_data` -> `bt_mesh_provision` -> `bt_mesh_start` -> `bt_mesh_pb_gatt_disable` -> `is_register = false`. Then later `pb_gatt.c:gatt_disconnect` gets called, but the all is dropped because `is_register` is false.

LingaoM commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -369,10 +353,17 @@ int bt_mesh_start(void)
-		bt_mesh_proxy_gatt_enable();
-		bt_mesh_adv_update();
+	/* For PB-GATT provision, will enable in le disconnect handler. */
+	if (bt_mesh_prov_link.bearer->type == BT_MESH_PROV_ADV) {
```
Only occurs when self provision or restore?

PavelVPV replied:
As far as I see, `bt_mesh_prov_link.bearer` is initialized only when the provisioning link is open. So, it isn't initialized at boot-up.

LingaoM committed:
Bluetooth: Mesh: Remove unnecessary prov buf get function

Remove `bt_mesh_pb_gatt_get_buf`.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -370,10 +355,16 @@ int bt_mesh_start(void)
+		}
+	} else if (bt_mesh_prov_link.bearer->type == BT_MESH_PROV_ADV) {
+		if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
+			(void)bt_mesh_pb_gatt_disable();
```
How does Proxy server gets enabled if device was provisioned through PB-ADV?

PavelVPV commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -370,10 +355,17 @@ int bt_mesh_start(void)
+		}
+
+		if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY) &&
+		    bt_mesh_gatt_proxy_get() != BT_MESH_GATT_PROXY_NOT_SUPPORTED) {
```
Minor: `bt_mesh_gatt_proxy_get` already checks for `IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)`. So `if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY))` should be enough.

LingaoM commented:
@trond-snekvik Can you take a moment to review it?

trond-snekvik commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,348 @@
+	},
+};
+
+static bool is_register;
```
```suggestion
static bool service_registered;
```

trond-snekvik commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,348 @@
+static uint8_t __noinit client_buf_data[CLIENT_BUF_SIZE];
+static struct bt_mesh_proxy_role cli = {
+	.cb.send = gatt_send,
+	.buf = {
+		.__buf = client_buf_data,
+		.data  = client_buf_data,
+		.size  = CLIENT_BUF_SIZE,
+		.len   = CLIENT_BUF_SIZE,
+	},
```
Now that this buffer is being instantiated in more places, it seems like it'd be better to just put it in proxy_msg.c, so we don't need to allocate 66 + 30 * conn_count bytes for this. The buffer operation is happening in that file anyway, and there's no scenario where we would run more than `conn_count` number of parallel buffers, regardless of roles.

With this change, I think it'd make sense to move the common part of the gatt connect/disconnect callbacks to that file as well, as they're doing the same thing in both files. If that sits in proxy_msg.c, we only need to handle the "add proxy service" part of the pb_gatt callback.

LingaoM replied:
but, how define buf in proxy_msg, there are different length between pbgatt and proxy.

LingaoM replied:
Good idea for me👍

trond-snekvik replied:
> but, how define buf in proxy_msg, there are different length between pbgatt and proxy.

perhaps this needs to be `#ifdef` wrapped? ie if GATT proxy is available, it's 30 * conn_count , else it's 66. Could also use a max

LingaoM replied:
I think it should be. It seems unrealistic to completely eliminate config in this file, but it still makes sense to move proxy_complete_msg to the corresponding role.

LingaoM replied:
One detail needs to be clarified. I think for pb-gatt, only one conn is needed. Therefore, max may be used for the first buf, and 30 for the latter?

LingaoM replied:
Before putting conn_cb in the rom section, it is really friendly to our current approach, the structure is very flexible (at least I think it is), and it also saves some RAM resources.

LingaoM replied:
In trying this scheme, I found that proxy_msg does not have the ability to obtain client (defined in proxy_srv.c) and cli (defined in pb_gatt_srv.c).

trond-snekvik replied:
I suppose parts of the role structure would have to move as well?

LingaoM replied:
how to do?

LingaoM replied:
If we understand correctly, do you want to define the role structure in the proxy_msg.c, referenced by proxy_srv.c or pb_gatt_srv.c?

LingaoM replied:
``` C
static struct bt_mesh_proxy_client {
	struct bt_mesh_proxy_role *cli;
	uint16_t filter[CONFIG_BT_MESH_PROXY_FILTER_SIZE];
	enum __packed {
		NONE,
		WHITELIST,
		BLACKLIST,

```

this ?

trond-snekvik replied:
Seems like a sensible solution to me, but your call

LingaoM replied:
Done.

trond-snekvik commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,348 @@
+
+	bt_mesh_pb_gatt_close(conn);
+
+	if (bt_mesh_is_provisioned()) {
+		(void)bt_mesh_pb_gatt_disable();
+
+		if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY) &&
+		    bt_mesh_gatt_proxy_get() !=
+			BT_MESH_GATT_PROXY_NOT_SUPPORTED) {
+			(void)bt_mesh_proxy_gatt_enable();
+		}
+	}
```
This could be moved to the proxy server's callback, I think. It'll need to treat the `info.role != BT_CONN_ROLE_SLAVE|| !is_register` part a little different, but it'll eliminate yet another `IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)` check. It'll also remove the need for ordering the callbacks (I think?)

LingaoM replied:
I still think it can be used temporarily, you can modify it at any time if you have a better idea later.

LingaoM replied:
Done.

trond-snekvik commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -69,16 +58,9 @@
 	.interval_min = BT_GAP_ADV_FAST_INT_MIN_2,                             \
 	.interval_max = BT_GAP_ADV_FAST_INT_MAX_2
```
These `ADV_` defines could be moved to proxy.h, I believe. They're duplicated in proxy_srv.c and pb_gatt.c, but in pb_gatt.c, they're handling `CONFIG_BT_MESH_PROXY_USE_DEVICE_NAME` wrong

trond-snekvik commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -90,34 +72,21 @@ static struct bt_mesh_proxy_client {
 		NONE,
 		WHITELIST,
 		BLACKLIST,
-		PROV,
 	} filter_type;
-#if defined(CONFIG_BT_MESH_GATT_PROXY)
 	struct k_work send_beacons;
-#endif /* CONFIG_BT_MESH_GATT_PROXY */
 } clients[CONFIG_BT_MAX_CONN] = {
 	[0 ... (CONFIG_BT_MAX_CONN - 1)] = {
-#if defined(CONFIG_BT_MESH_GATT_PROXY)
 		.send_beacons = Z_WORK_INITIALIZER(proxy_send_beacons),
-#endif /* CONFIG_BT_MESH_GATT_PROXY */
 		.cli.cb = {
 			.send = proxy_send,
-#if defined(CONFIG_BT_MESH_GATT_PROXY)
 			.recv = proxy_filter_recv,
-#endif /* CONFIG_BT_MESH_GATT_PROXY */
```
Looks much cleaner :star:

trond-snekvik commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -90,34 +72,21 @@ static struct bt_mesh_proxy_client {
 		.cli.cb = {
 			.send = proxy_send,
-#if defined(CONFIG_BT_MESH_GATT_PROXY)
 			.recv = proxy_filter_recv,
```
I think we discussed this in the previous PR as well, but IMO, it'd make sense to replace proxy_msg.c's `proxy_complete_pdu()` with this callback. This'd remove the last `IS_ENABLED()` macros from the common file, and wouldn't introduce any duplicate code at all.

I.e., instead of calling `proxy_complete_pdu()`, `bt_mesh_proxy_msg_recv()` would call this callback immediately, with the role and buffer pointers.

LingaoM replied:
Look like make sense to me 👍

trond-snekvik commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -899,250 +855,45 @@ static void gatt_connected(struct bt_conn *conn, uint8_t err)
+	/* If this fails, the work handler exits early, as
+	 * there's no active connection.
+	 */
+	(void)k_work_cancel_delayable(&client->cli.sar_timer);
```
Shouldn't this (and the corresponding `bt_mesh_proxy_msg_init`) be called in the pb_gatt.c file as well? I believe that requires proxy SAR too? 
If you follow the suggestion to move cli.buf into proxy_msg.c, we could probably move this too to that file.

LingaoM replied:
I'm think call this on gatt connect cb handler will be better, also will alloc buf in this function too.

trond-snekvik commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,348 @@
+	return err;
+}
+
+/* Add `_2` suffix forces proxy conn_cb to precede pb_gatt conn_cb.
+ * As we may register proxy services in pb_gatt disconnect cb.
+ */
```
```suggestion
/* Add `_2` suffix forces this callback to be called after proxy_srv's conn_cb.
 */
```

trond-snekvik reviewed (changes requested):
Sorry for the delayed review.

Looks much cleaner already! I have a few suggestions, mostly triggered by the disconnect callback ordering and some placement of responsibility between proxy_msg.c and the two role specific modules.

LingaoM committed:
Bluetooth: Mesh: Move command buffer to proxy_msg.c

Move command buffer alloc to proxy_msg.c.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Move proxy complete message to seperate role

Move proxy complete message to seperate role.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Move bt_mesh_proxy_role to proxy_msg.c

Mesh bt_mesh_proxy_role structure to proxy_msg.c

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Reflact disconnect logic to separate roles

remove section ordering to use code more readable.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik commented on subsys/bluetooth/mesh/proxy.h:
```diff
@@ -4,38 +4,46 @@
+#define ADV_OPT_PROV                                                           \
+	(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_SCANNABLE |                 \
+	 BT_LE_ADV_OPT_ONE_TIME | ADV_OPT_USE_IDENTITY |                       \
+	 BT_LE_ADV_OPT_USE_NAME)
```
@PavelVPV introduced `CONFIG_BT_MESH_PROXY_USE_DEVICE_NAME` IIRC. Do you think we could/should use that to configure the provisioning advertisement as well?

PavelVPV replied:
It should have a different config option for sure.

LingaoM replied:
Done.

trond-snekvik commented on subsys/bluetooth/mesh/proxy_msg.c:
```diff
@@ -243,7 +190,70 @@ int bt_mesh_proxy_msg_send(struct bt_mesh_proxy_role *role, uint8_t type,
+{
+	struct bt_mesh_proxy_role *role;
+
+	role = &roles[bt_conn_index(conn)];
```
The buffer sizes assume that we'll only use `role[0]` when running PB-GATT, but if there already were another connection open when the PB-GATT connection was established, it'll be `1` here. What will happen to the buffer in this case?

LingaoM replied:
Isn't the connection using pb-gatt 0?

LingaoM replied:
Do you mean that mesh may coexist with other samples?

LingaoM replied:
If this is the case, then I have to define all of them as 66 bytes, which is consistent with now.

LingaoM replied:
I'm fine with all connection use same buffer len, and it's true for pb-gatt client.
So, add this now is OK.

trond-snekvik commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,306 @@
+
+	bt_conn_get_info(conn, &info);
+	if (info.role != BT_CONN_ROLE_SLAVE ||
+	    !service_registered) {
```
Should also check for `bt_mesh_is_provisioned()` to avoid race conditions on the callbacks, no?

LingaoM replied:
Already set to true in bt_mesh_provision.

LingaoM replied:
BT_MESH_VALID

LingaoM replied:
Done.

trond-snekvik commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,306 @@
+		break;
+	}
+
+	net_buf_simple_reset(&role->buf);
```
Should be done by proxy_msg IMO.

LingaoM replied:
Keep consistent.

trond-snekvik reviewed (changes requested):
Looks pretty clean now, but I have some concerns about the changes in this last push:

LingaoM committed:
Bluetooth: Mesh: Add seperate config for pb-gatt dev name

Add seperate config to control whether device name include in
pb-gatt advertising scan response data.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Use common buf size for all conn

Use common buffer size for all bluetooth mesh connection.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/pb_gatt_srv.c:
```diff
@@ -0,0 +1,315 @@
+#define ADV_OPT_PROV                                                           \
+	(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_SCANNABLE |                 \
+	 BT_LE_ADV_OPT_ONE_TIME | ADV_OPT_USE_IDENTITY |                       \
+	 BT_LE_ADV_OPT_USE_NAME)
```
```suggestion
	 ADV_OPT_USE_NAME)
```

LingaoM replied:
This is an existing problem, I did not notice.

LingaoM commented on subsys/bluetooth/mesh/proxy_msg.c:
```diff
@@ -48,6 +48,19 @@
+#define PROXY_MSG_BUF_LEN PB_GATT_BUF_LEN_MAX
+#else
+#define PROXY_MSG_BUF_LEN PROXY_BUF_LEN_MAX
+#endif
```
Maybe move this to kconfig better, so that we can all del config in this file.

LingaoM committed:
Bluetooth: Mesh: Move proxy message size to kconfig

Add `BT_MESH_PROXY_MSG_LEN` to config proxy message length

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -120,6 +126,13 @@ config BT_MESH_PB_GATT
 
+config BT_MESH_PB_GATT_USE_DEVICE_NAME
+	bool "Include Bluetooth device name in scan response"
+	depends on BT_MESH_PB_GATT
```
Should this be enabled by default to keep the old behavior?

LingaoM commented:
Do we plan to merge 2.7 or wait?

cfriedt commented:
@jhedberg - would you like to ack?

cfriedt commented:
@jhedberg - can you also ack?

LingaoM commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -824,26 +801,21 @@ bool bt_mesh_proxy_relay(struct net_buf *buf, uint16_t dst)
 			continue;
 		}
 
-		if (client->filter_type == PROV) {
```
https://github.com/zephyrproject-rtos/zephyr/pull/40158#issuecomment-963981835

@jhedberg
